### PR TITLE
Markdown 3.1.2 Release

### PR DESCRIPTION
**Pull request by cmcnally-r7**
* Improved error messaging in markdown_to_txt

* Refactored html to markdown

* Refactored html & markdown

* Removed empty variable in markdown_to_txt

* Update version and help.md

* Update help.md

* Added unit test for markdown_to_txt

* Unit test templates

* Unit test filler

* Reformatted to_txt with parameterized

* 3/4 unit tests finished

* Tests finished minus to pdf

* Reformatting

* Fix static code check

* Longshot to fix the pandoc error

* Second longshot to fix the pandoc error

* Finished up on unit tests

* Tided up markdowntopdf test

* Bug fix for markdown to txt

* Updated help md for bug fix

* Plugin validator fix

### Description

Describe the proposed changes:

  -

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/markdown/help.md [hunk 1]
```diff
@@ -180,6 +184,7 @@ _This plugin does not contain any troubleshooting information._
 
 # Version History
 
+* 3.1.2 - Added additional error messaging | Refactored code | Fixed bug in action
```
Comment (ablakley-r7) [inline plugins/markdown/help.md:187]:
```suggestion
* 3.1.2 - Added additional error messaging | Refactored code | Fixed bug in Markdown to TXT action which resulted in an incorrect output
```

File: plugins/markdown/komand_markdown/actions/markdown_to_pdf/action.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 def makePDF(html: str, path: str) -> str:
     infile = path + "str.html"
     outfile = path + "tmp.pdf"
-    with open(infile, "w") as f:
+    with open(infile, "w", encoding="utf-8") as f:
```
Comment (igorski-r7) [inline plugins/markdown/komand_markdown/actions/markdown_to_pdf/action.py:13]:
```suggestion
    with open(infile, "w", encoding="utf-8") as file_:
```

File: plugins/markdown/komand_markdown/actions/markdown_to_html/action.py [hunk 3]
```diff
@@ -13,13 +14,14 @@ def __init__(self):
...
+        if not (((instr is None) ^ (inbytes is None)) or ((instr == "") ^ (inbytes == ""))):
+            raise PluginException(
+                cause="Input error",
+                assistance="Only one of Markdown or Markdown String can be defined"
                 if instr != inbytes
-                else "You must define one of Markdown or Markdown String."
+                else "You must define one of Markdown or Markdown String.",
             )
         if instr:
             html_string = utils.convert(instr, "md", "html")
```
Comment (ablakley-r7) [inline plugins/markdown/komand_markdown/actions/markdown_to_html/action.py:27]: Possible to fix end of this run method to be similar to Markdown to HTML action?

File: plugins/markdown/komand_markdown/actions/markdown_to_pdf/action.py (hunk 2 above)
Comment (cmcnally-r7) [reply plugins/markdown/komand_markdown/actions/markdown_to_pdf/action.py:13]: Would it be `file_` or `file`? 
Does the underscore mean it is an unused variable?

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]

File: plugins/markdown/komand_markdown/actions/markdown_to_html/action.py (hunk 3 above)
Comment (cmcnally-r7) [reply plugins/markdown/komand_markdown/actions/markdown_to_html/action.py:27]: How do you mean? This is the markdown to html action - do you mean a different one?

**Review by igorski-r7 — APPROVED** [review]

**Review by ablakley-r7 — APPROVED** [review]

**Review by blaxminarayan-r7 — APPROVED** [review]

**Pull request merged**